### PR TITLE
Improvement Upload CV Bill Account

### LINKED ISSUE #291: Improvement Upload CV Bill Account
image

### PR DESCRIPTION
**Pull request by khaletuanncc**
https://github.com/ncc-erp/ncc-erp-project/issues/291

## Release Notes

- **New Features**
  - Introduced functionality for users to upload and download CVs associated with project accounts.
  - Added a new component for managing CV uploads, enhancing user interaction with file handling.
  - Implemented visual improvements, including hover effects and responsive design adjustments.

- **Bug Fixes**
  - Improved error handling during file uploads to enhance user experience.

- **Documentation**
  - Updated relevant documentation to reflect new features and usage instructions.

- **Tests**
  - Added test suite for the new CV upload component to ensure functionality and reliability.

### CODE REVIEW

File: angular/src/app/service/model/upload-cv.dto.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export class GetCvBillAccountDto {
+    id: number;
+    nameCv: string;
+    linkCV: string;
+}
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/service/model/upload-cv.dto.ts]: unnecessary

File: aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/Dto/CvBillAccountDto.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace ProjectManagement.Services.ProjectUserBill.Dto
+{
+    public class UploadCvBillAccountDto : EntityDto<long>
+    {
+        [Required]
+        public string NameCv { get; set; }
+        [Required]
+        public IFormFile SelectedFile { get; set; }
+    }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/Dto/CvBillAccountDto.cs]: unnecessary

File: angular/src/app/service/utility/file-handler.service.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+})
+export class FileHandlerService {
+
+  constructor() { }
+
+  downloadFile(base64Data: string, fileName: string, mimeType: string = 'application/octet-stream'): void {
+    const byteArray = this.s2ab(atob(base64Data));
+    const fileBlob = new Blob([byteArray], { type: mimeType });
+    FileSaver.saveAs(fileBlob, fileName);
+  }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/service/utility/file-handler.service.ts:15]: unnecessary

File: angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UploadCvBillAccountComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.spec.ts:24]: unnecessary

File: angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<div class="all_dialog">
+    <h3 class="header_h h_drag_and_drop">Upload CV</h3>
+    <form action="" class="py-2 border-top">
+        <div class="d-flex flex-column">
+            <label for="file">Choose file:</label>
+            <input type="file"
+                class="form-group"
+                id="file"
+                required
+                (change)="handleFileInput($event)">
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.html]: ok

File: angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        <div class="d-flex flex-column">
+            <mat-form-field appearance="outline" floatLabel="auto">
+                <mat-label>Enter Name CV</mat-label>
+                <input name="nameCv"
+                    type="text"
+                    matInput
+                    required
+                    [(ngModel)]="billAccountCv.nameCv"
+                />
+            </mat-form-field>
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.html]: unnecessary

File: aspnet-core/src/ProjectManagement.Core/Entities/ProjectUserBill.cs [hunk 7]
```diff
@@ -32,7 +32,7 @@ public class ProjectUserBill : FullAuditedEntity<long>, IMayHaveTenant
         public bool isExpose { get; set; }
         public bool isActive { get; set; }
         public ChargeType? ChargeType { get; set; }
-
+        public string NameCv { get; set; }
         public string LinkCV { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Entities/ProjectUserBill.cs]: unnecessary

File: angular/src/app/service/model/bill-info.interface.ts [hunk 8]
```diff
@@ -50,6 +50,8 @@ export interface IProject {
     linkedResources: ILinkedResource[];
     userSkills: IUserSkill[];
     skillNote: string;
+    linkCV: string;
+    nameCv: string;
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/service/model/bill-info.interface.ts]: unnecessary

File: angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      .pipe(catchError(error => this.projectUserBillService.handleError(error)))
+      .subscribe({
+        next: res => {
+          abp.notify.success("Upload CV successfully");
+          this.billAccountCv = res.result;
+          this.dialogRef.close(this.billAccountCv);
+        },
+        error: () => { this.dialogRef.close() }
+      });
+  }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.ts]: have handled in different function

File: angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    this.billAccountCv = this.data;
+  }
+
+  handleFileInput(event: Event): void {
+    const input = event.target as HTMLInputElement;
+    this.selectedFile = input?.files?.item(0) || null;
+    if (this.selectedFile && !this.billAccountCv.nameCv) {
+      this.billAccountCv.nameCv = this.selectedFile.name.replace(/\.[^/.]+$/, "");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/components/upload-cv-bill-account/upload-cv-bill-account.component.ts]: ok

File: aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/Dto/GetProjectUserBillDto.cs [hunk 11]
```diff
@@ -51,6 +51,7 @@ public class GetProjectUserBillDto : EntityDto<long>
         public List<GetUserInfo> LinkedResources { get; set; }
         public float totalHeadCount { get; set; }
         public GridParam GridParam { get; set; }
+        public string NameCv { get; set; }
         public string LinkCV { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/Dto/GetProjectUserBillDto.cs]: unnecessary

File: angular/src/app/service/api/project-user-bill.service.ts [hunk 12]
```diff
@@ -117,5 +118,12 @@ export class ProjectUserBillService extends BaseApiService {
     return this.http.post<any>(this.rootUrl + '/UpdateBillUserSkill', request);
   }
 
+  UploadCvBillAccount(request: UploadCvBillAccountDto): Observable<any>{
+    const formData = new FormData();
+    formData.append('Id', request.id.toString());
+    formData.append('NameCv', request.nameCv);
+    formData.append('SelectedFile', request.selectedFile);
+    return this.http.post<any>(this.rootUrl + '/UploadCvBillAccount', formData);
+  }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/service/api/project-user-bill.service.ts]: unnecessary

File: angular/src/shared/shared.module.ts [hunk 13]
```diff
@@ -164,7 +165,8 @@ import { MultipleSelectComponent } from './components/multiple-select/multiple-s
         SelectClearOptionComponent,
         LineBreakPipe,
         MultipleSelectStringValueComponent,
-        MultipleSelectComponent
+        MultipleSelectComponent,
+        UploadCvBillAccountComponent
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/shared/shared.module.ts:169]: unnecessary

File: angular/src/app/modules/pm-management/list-project/list-project-detail/project-bill/project-bill.component.ts [hunk 14]
```diff
@@ -519,35 +522,23 @@ export class ProjectBillComponent extends AppComponentBase implements OnInit {
...
       this.projectUserBillService.DownloadCVLink(id).subscribe(data => {
-        const file = new Blob([this.s2ab(atob(data.result.data))],{
-          type: "application/vnd.ms-excel;charset=utf-8"
-          
-        });
-        FileSaver.saveAs(file,data.result.fileName);
+        this.fileHandlerService.downloadFile(data.result.data, data.result.fileName);
       })
       window.open('_blank');
     }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/pm-management/list-project/list-project-detail/project-bill/project-bill.component.ts:539]: unnecessary

File: angular/src/app/modules/delivery-management/delivery/available-resource-tab/bill-account-plan/bill-account-plan.component.ts [hunk 15]
```diff
@@ -392,4 +395,23 @@ export class BillAccountPlanComponent
...
+        this.billInfoList[itemIndex].projects[projectIndex] = updatedProject;
+      }
+    });
+  }
+
+  downloadFile(id: number){
+    this.projectUserBillService.DownloadCVLink(id).subscribe(data => {
+      this.fileHandlerService.downloadFile(data.result.data, data.result.fileName);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/delivery-management/delivery/available-resource-tab/bill-account-plan/bill-account-plan.component.ts:416]: browser auto handle this

File: aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/ProjectUserBillManager.cs [hunk 16]
```diff
@@ -498,5 +502,24 @@ private void ValidateProjectUserBill(long projectUserBillId)
...
+            }
+            projectUserBill.LinkCV = filePath;
+            await _workScope.UpdateAsync(projectUserBill);
+            return new GetCvBillAccountDto()
+            {
+                Id = projectUserBill.Id,
+                NameCv = projectUserBill.NameCv,
+                LinkCV = projectUserBill.LinkCV
+            };
+        }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ProjectUserBill/ProjectUserBillManager.cs]: handle in frontend side

**Review by tiennguyenhuuncc — APPROVED** [review]

**Pull request merged**